### PR TITLE
Sync: Product Pixels

### PR DESCRIPTION
**Pull request by malmstein**
Task/Issue URL: https://app.asana.com/0/488551667048375/1206203109961669/f 

### Description
This PR adds the required Product Pixels for the soft release.
Note: Duck address override pixel was already implemented

### Steps to test this PR

_Success rate daily_
You’ll need to run some sync operations and force some api errors. 
In `SyncDailyReportingWorkerScheduler`, comment out the enqueuing in the WorkManager
In `RealSyncEngine`, comment out line 115
- [x] Fresh install, enable sync
- [x] Add some bookmarks and folders
Force some API errors as described in the task
- [x] Ensure errors are stored in the DB
In SyncStatsRepository, change the date filter to be today and not yesterday 
- [x] Ensure errors are stored in the DB
Remove the comment in `SyncDailyReportingWorkerScheduler` so that the work is enqueued and install the app again
Remove the comment in `RealSyncEngine` line 115
- [x] Verify pixel is sent

_Timestamp resolution triggered_
Pixel sent when there are local changes after a sync operation has started
- [ ] Device A -> Add a bookmark and enable sync
- [ ] Device B -> Join sync account from Device A
- [ ] Device A -> Open bookmark edit dialog
- [ ] Device B -> Rename bookmark
- [ ] Device A -> Rename bookmark
- [ ] Device A ->  Verify timestamp resolution conflict triggered, bookmark has name from Device A

### CODE REVIEW

**Comment by malmstein** [comment]
<b>In this stack:</b> <code>🔄 Sync product pixels, improve tracking and analytics.</code><a href="https://withgraphite.notion.site/Graphite-beta-features-EXTERNAL-496843566b204c7a97a2a21f8dea5434">(?)</a>

* `develop`
  * **#4074** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/4074?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @malmstein and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/engine/SyncApiClient.kt [hunk 1]
```diff
@@ -83,70 +84,26 @@ class AppSyncApiClient @Inject constructor(
...
-                        Result.Error(result.code, result.reason)
-                    }
-                }
-            }
-
-            is Result.Success -> {
-                val remoteChanges = mapResponse(type, result.data)
-                Result.Success(remoteChanges)
-            }
+        val result = when (type) {
```
Comment (cmonfortep) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/engine/SyncApiClient.kt:95]: nice clean up here 👍

**Review by cmonfortep — COMMENTED** [review]
Looks good.

Tested first scenario, and works well.

Left some comments for us to review, and it's pending to clarify the timestamp ones as you mentioned when you are back tomorrow. Thanks!

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/error/SyncApiErrorReporter.kt [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import com.duckduckgo.sync.impl.Result.Error
+import com.duckduckgo.sync.impl.pixels.SyncPixels
+import com.duckduckgo.sync.store.model.SyncApiErrorType.OBJECT_LIMIT_EXCEEDED
+import com.duckduckgo.sync.store.model.SyncApiErrorType.REQUEST_SIZE_LIMIT_EXCEEDED
+import com.duckduckgo.sync.store.model.SyncApiErrorType.TOO_MANY_REQUESTS
+import com.duckduckgo.sync.store.model.SyncApiErrorType.VALIDATION_ERROR
+import com.squareup.anvil.annotations.ContributesBinding
+import javax.inject.Inject
+
+interface SyncApiErrorReporter {
```
[the remark this answers is not in the document]
Comment (malmstein) [reply sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/error/SyncApiErrorReporter.kt:31]: I like that `SyncApiErrorRecorder`

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/error/SyncApiErrorRepository.kt [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        if (todaysError == null) {
+            apiErrorDao.insert(SyncApiError(feature = feature.field, errorType = apiError.name, count = 1, date = today))
+        } else {
+            apiErrorDao.incrementCount(feature.field, apiError.name, today)
+        }
+    }
+
+    override fun getErrorsByDate(date: String): List<SyncApiErrorPixelData> {
+        return apiErrorDao.errorsByDate(date).map {
+            val errorType = when (it.errorType) {
```
[the remark this answers is not in the document]
Comment (malmstein) [reply sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/error/SyncApiErrorRepository.kt:58]: I disagree. `SyncApiErrorType` represents the different values that error can have. It has nothing to do with Pixels, not it should do. This is exactly why Repostitories exist, return a data in the desired format, mapping from db models to domain models. 

Adding this logic to the enum leaks the implementation detail of another class.

File: sync/sync-impl/src/test/java/com/duckduckgo/sync/impl/engine/SyncApiClientTest.kt [hunk 4]
```diff
@@ -50,7 +49,7 @@ internal class SyncApiClientTest {
 
     @Before
     fun before() {
-        apiClient = AppSyncApiClient(syncStore, syncApi, syncPixels)
+        apiClient = AppSyncApiClient(syncStore, syncApi, syncApiErrorReporter)
```
[the remark this answers is not in the document]
Comment (malmstein) [reply sync/sync-impl/src/test/java/com/duckduckgo/sync/impl/engine/SyncApiClientTest.kt:52]: agreed, I’ve added tests for those cases

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/RealSyncCrypto.kt [hunk 5]
```diff
@@ -28,12 +27,10 @@ import javax.inject.Inject
 class RealSyncCrypto @Inject constructor(
     private val nativeLib: SyncLib,
     private val syncStore: SyncStore,
-    private val syncPixels: SyncPixels,
 ) : SyncCrypto {
     override fun encrypt(text: String): String {
         val encryptResult = nativeLib.encryptData(text, syncStore.secretKey.orEmpty())
         return if (encryptResult.result != 0L) {
```
[the remark this answers is not in the document]
Comment (malmstein) [reply sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/RealSyncCrypto.kt:33]: We can, not as part of this PR. See https://app.asana.com/0/1201493110486074/1205581521828305

File: sync/sync-store/src/main/java/com/duckduckgo/sync/store/model/SyncDatabaseModels.kt [hunk 6]
```diff
@@ -34,10 +34,31 @@ data class SyncAttempt(
...
     FAIL,
 }
+
+@Entity(
+    tableName = "sync_api_errors",
+)
+data class SyncApiError(
+    @PrimaryKey(autoGenerate = true) val id: Int = 0,
+    val feature: String,
+    val errorType: String,
```
[the remark this answers is not in the document]
Comment (malmstein) [reply sync/sync-store/src/main/java/com/duckduckgo/sync/store/model/SyncDatabaseModels.kt]: Yeah, I like that. Moving to the TypeConverters solution

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/pixels/SyncPixels.kt [hunk 7]
```diff
@@ -217,26 +154,22 @@ class RealSyncPixels @Inject constructor(
 
 // https://app.asana.com/0/72649045549333/1205649300615861
 enum class SyncPixelName(override val pixelName: String) : Pixel.PixelName {
-    SYNC_DAILY_PIXEL("m_sync_daily"),
+    SYNC_DAILY_PIXEL("m_sync_success_rate_daily"),
+    SYNC_ORPHAN_PRESENT("m_sync_%s_orphan_present"),
```
[the remark this answers is not in the document]
Comment (malmstein) [reply sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/pixels/SyncPixels.kt]: Good catch, forgot to remove it. I’m getting rid of it and we can discuss if we really need it.

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt [hunk 8]
```diff
@@ -71,12 +77,12 @@ class CredentialsLastModifiedWinsStrategy(
         localCredential: LoginCredentials,
         remoteEntry: CredentialsSyncEntryResponse,
         clientModifiedSince: String,
-    ) {
+    ): Boolean {
```
[the remark this answers is not in the document]
Comment (malmstein) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt:80]: I’ve added a description to the method, I prefer not to rename it

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/engine/RealSyncEngine.kt [hunk 9]
```diff
@@ -227,11 +225,16 @@ class RealSyncEngine @Inject constructor(
             when (val result = it.onSuccess(remoteChanges, conflictResolution)) {
                 is SyncMergeResult.Success -> {
                     if (result.orphans) {
-                        syncPixels.fireOrphanPresentPixel(remoteChanges.type.toString())
+                        Timber.d("Sync - Orphans present in this sync operation for feature ${remoteChanges.type.field}")
+                        syncPixels.fireOrphanPresentPixel(remoteChanges.type.field)
+                    }
+                    if (result.timestampConflict) {
```
[the remark this answers is not in the document]
Comment (malmstein) [reply sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/engine/RealSyncEngine.kt:230]: yes, done

File: saved-sites/saved-sites-impl/src/main/java/com/duckduckgo/savedsites/impl/sync/algorithm/SavedSitesSyncPersisterAlgorithm.kt [hunk 10]
```diff
@@ -80,27 +81,39 @@ class RealSavedSitesSyncPersisterAlgorithm @Inject constructor(
         // check all children, the ones that are not in allFolders don't have a parent
         val foldersWithoutParent = allFolderIds.filterNot { allChildren.contains(it) }
         foldersWithoutParent.forEach { folderId ->
-            if (savedSitesRepository.getFolder(folderId) != null) {
+            val folder = savedSitesRepository.getFolder(folderId)
+            if (folder != null) {
                 processIds.add(folderId)
             }
-            processFolder(folderId, SavedSitesNames.BOOKMARKS_ROOT, bookmarks.entries, lastModified, processIds, conflictResolution)
+            if (folder != null && folder.modifiedAfter(clientModifiedSince) && conflictResolution != DEDUPLICATION) {
```
[the remark this answers is not in the document]
Comment (malmstein) [reply saved-sites/saved-sites-impl/src/main/java/com/duckduckgo/savedsites/impl/sync/algorithm/SavedSitesSyncPersisterAlgorithm.kt]: these functions are recursive, I can’t check for this condition inside because otherwise the values won’t be correct.
we check for deletion afterwards, not here. 
we can probably make the algorithm more readable, but it’s not something I’d like to do now.

**Comment by malmstein** [comment]
@cmonfortep PR comments addressed. Check the main task for the conversation around the timestamp conflict.

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/pixels/SyncPixels.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (cmonfortep) [reply sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/pixels/SyncPixels.kt]: If I'm not wrong, orphans are now measured in the BE (https://app.asana.com/0/0/1205649300615861/f), so it's fine to remove 👍

File: sync/sync-store/src/main/java/com/duckduckgo/sync/store/SyncDatabase.kt [hunk 11]
```diff
@@ -51,8 +56,21 @@ object SyncTypeConverters {
...
+    fun toSyncApiErrorType(errorType: String): SyncApiErrorType {
+        return try {
+            SyncApiErrorType.valueOf(errorType)
+        } catch (t: Throwable) {
+            SyncApiErrorType.TOO_MANY_REQUESTS
+        }
+    }
+
+    @TypeConverter
+    fun fromSSyncApiErrorType(errorType: SyncApiErrorType): String {
```
Comment (cmonfortep) [inline sync/sync-store/src/main/java/com/duckduckgo/sync/store/SyncDatabase.kt:73]: NIT: typo in the name

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/error/SyncApiErrorRepository.kt [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+                SyncApiErrorType.TOO_MANY_REQUESTS -> SyncPixelParameters.TOO_MANY_REQUESTS
+                SyncApiErrorType.OBJECT_LIMIT_EXCEEDED -> SyncPixelParameters.OBJECT_LIMIT_EXCEEDED_COUNT
+                SyncApiErrorType.REQUEST_SIZE_LIMIT_EXCEEDED -> SyncPixelParameters.REQUEST_SIZE_LIMIT_EXCEEDED_COUNT
+                SyncApiErrorType.VALIDATION_ERROR -> SyncPixelParameters.VALIDATION_ERROR_COUNT
+                else -> {
```
Comment (cmonfortep) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/error/SyncApiErrorRepository.kt:63]: do we need an `else` here with `when`+ `enum`?

File: sync/sync-store/src/main/java/com/duckduckgo/sync/store/SyncDatabase.kt [hunk 13]
```diff
@@ -51,8 +56,21 @@ object SyncTypeConverters {
...
     fun fromSyncState(syncState: SyncAttemptState): String {
         return syncState.name
     }
+
+    @TypeConverter
+    fun toSyncApiErrorType(errorType: String): SyncApiErrorType {
+        return try {
+            SyncApiErrorType.valueOf(errorType)
+        } catch (t: Throwable) {
+            SyncApiErrorType.TOO_MANY_REQUESTS
```
Comment (cmonfortep) [inline sync/sync-store/src/main/java/com/duckduckgo/sync/store/SyncDatabase.kt:68]: Won't be true to signal a too many request here. what about having `unknown` as fallback type? and if `unknown`, we don't send a pixel.

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (cmonfortep) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt:85]: Sorry for being a pain with this detail, but I don't like the method returning a `Boolean` that represents an implementation detail of the method.

The I read:
```
if (processExistingEntry(localCredential, entry, clientModifiedSince)) {
    hasDataChangedWhileSyncing = true
}
```
and I don't see why when `processExistingEntry` is `true` means `hasDataChangedWhileSyncing` is `true`. So it's hard to follow.

I see 2 options:
a) Return `SyncMergeResult` instead (this would be my pref)
b) change the definition of that `Boolean` to be `true if remote entry wins, false otherwise`.

File: saved-sites/saved-sites-impl/src/main/java/com/duckduckgo/savedsites/impl/sync/algorithm/SavedSitesSyncPersisterAlgorithm.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (cmonfortep) [reply saved-sites/saved-sites-impl/src/main/java/com/duckduckgo/savedsites/impl/sync/algorithm/SavedSitesSyncPersisterAlgorithm.kt:88]: let's forget readability then.
At least I would say we need to be consistent with the pixel in other models, and if delete shouldn't trigger the pixel, we should not count it here, is that a correct assumption?

[the remark this answers is not in the document]
Comment (malmstein) [reply saved-sites/saved-sites-impl/src/main/java/com/duckduckgo/savedsites/impl/sync/algorithm/SavedSitesSyncPersisterAlgorithm.kt:88]: `processFolder` will never process a deleted item. A deleted item won’t trigger a pixel. What made you think it does? What is not consistent?

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt [hunk 14]
```diff
@@ -37,27 +37,43 @@ class CredentialsLastModifiedWinsStrategy(
...
                 credentials.entries.forEach { entry ->
                     val localCredential = credentialsSync.getCredentialWithSyncId(entry.id)
                     if (localCredential == null) {
                         processNewEntry(entry, clientModifiedSince)
                     } else {
-                        processExistingEntry(localCredential, entry, clientModifiedSince)
+                        val localId = localCredential.id!!
+                        if (entry.isDeleted()) {
+                            Timber.d("Sync-autofill-Persist: >>> delete local ${localCredential.id}")
+                            credentialsSync.deleteCredential(localId)
```
Comment (cmonfortep) [inline autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt:51]: We should `return` at this point. If entry deleted, rest of the logic should not be executed.

File: saved-sites/saved-sites-impl/src/main/java/com/duckduckgo/savedsites/impl/sync/algorithm/SavedSitesSyncPersisterAlgorithm.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (cmonfortep) [reply saved-sites/saved-sites-impl/src/main/java/com/duckduckgo/savedsites/impl/sync/algorithm/SavedSitesSyncPersisterAlgorithm.kt:88]: > In line 65 -> val allFolders = bookmarks.entries.filter { it.isFolder() } a deleted item won’t be returned here. processFolder will never process a deleted item.

I don't see how that is true. From what I see we process folders and bookmarks, including deleted ones, which is correct. Otherwise we wouldn't delete them locally if we don't process them.

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (malmstein) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt]: discussed offline, we are refactoring a little bit to clean it up

File: sync/sync-store/src/main/java/com/duckduckgo/sync/store/SyncDatabase.kt (hunk 13 above)
Comment (malmstein) [reply sync/sync-store/src/main/java/com/duckduckgo/sync/store/SyncDatabase.kt:68]: actually not, because the chain goes all the way. the right thing to do here is to throw IllegalArgumentException

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt (hunk 14 above)
Comment (malmstein) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt:51]: discussed offline

File: saved-sites/saved-sites-impl/src/main/java/com/duckduckgo/savedsites/impl/sync/algorithm/SavedSitesSyncPersisterAlgorithm.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (malmstein) [reply saved-sites/saved-sites-impl/src/main/java/com/duckduckgo/savedsites/impl/sync/algorithm/SavedSitesSyncPersisterAlgorithm.kt:88]: A deleted entity looks like this:
```
      {
        "deleted": "2023-05-23T11:35:01Z",
        "id": "2E085BF4-EB03-4B71-8848-5C1A30DCE53E",
        "last_modified": null
      },
```

If you look at the start of the function, you’ll see this:

```
val allFolders = bookmarks.entries.filter { it.isFolder() }
val allBookmarkIds = bookmarks.entries.filter { it.isBookmark() }.map { it.id }
```

Both of these won’t return deleted entities, because `isFolder` and `isBookmark` check for the `folder` and `page` field that are not present in deleted entities. For that reason, `processFolder` and `processBookmark` won’t iterate over deleted entities.

After we’ve processed non deleted items, we have this:

```
val allDeletedIds = bookmarks.entries.filter { it.deleted != null }.map { it.id }
processDeletedItems(allDeletedIds)

```
It’s here where we iterate over the deleted entities, and delete them. Why? Because it’s a tree structure, we can’t delete from the top. Subfolders might be deleted, but it’s content might have moved, we only know that after processing all the items in the payload. This is different in other models, where there’s no relation between entities. 

Going back to the original question.  We process deleted entities, and we don’t trigger the pixel for those entities.

File: sync/sync-store/src/main/java/com/duckduckgo/sync/store/SyncDatabase.kt (hunk 13 above)
Comment (cmonfortep) [reply sync/sync-store/src/main/java/com/duckduckgo/sync/store/SyncDatabase.kt]: That's correct. However, be careful if we ever remove an errorType or the db will crash when reading an old value.

File: saved-sites/saved-sites-impl/src/main/java/com/duckduckgo/savedsites/impl/sync/algorithm/SavedSitesSyncPersisterAlgorithm.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (cmonfortep) [reply saved-sites/saved-sites-impl/src/main/java/com/duckduckgo/savedsites/impl/sync/algorithm/SavedSitesSyncPersisterAlgorithm.kt:88]: gotcha! thanks for explaining ⭐️ 

I was under the assumption `isFolder` or `isBookmark` were not filtering deteled items, now it's clear.

**Review by cmonfortep — APPROVED** [review]
LGTM thanks for addressing and clarifying the comments.

The only small bit is the missing `return` in `CredentialsLastModifiedWinsStrategy` when an object is deleted.

Thanks for the work here.

File: autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt (hunk 14 above)
Comment (malmstein) [reply autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/sync/persister/CredentialsLastModifiedWinsStrategy.kt:51]: correct, adding the change now.
